xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -131,23 +137,129 @@ XdsClient createXdsClient() {
     };
     xdsClientPool = new RefCountedXdsClientObjectPool(xdsClientFactory);
     xdsClient = xdsClientPool.getObject();
-    xdsClient.watchConfigData(authority, new ConfigWatcher() {
-      @Override
-      public void onConfigChanged(ConfigUpdate update) {
+    xdsClient.watchConfigData(authority, new ConfigWatcherImpl(listener));
+  }
+
+  private class ConfigWatcherImpl implements ConfigWatcher {
+
+    final Listener2 listener;
+
+    ConfigWatcherImpl(Listener2 listener) {
+      this.listener = listener;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onConfigChanged(ConfigUpdate update) {
+      Map<String, ?> config;
+      if (update.getRoutes().size() > 1) {
+        // Path matching is enabled and the config update is not a single route. Generate
+        // xds-routing lb config.
+        logger.log(
+            XdsLogLevel.INFO,
+            "Received config update with {0} routes from xDS client {1}",
+            update.getRoutes().size(),
+            xdsClient);
+        logger.log(
+            XdsLogLevel.DEBUG,
+            "Received config update from xDS client {0}: {1}",
+            xdsClient,
+            update);
+        List<Object> routes = new ArrayList<>(update.getRoutes().size());
```

Comment:
This whole logic of generating LB configs is complicated, given that each LB config can have recursively nested child policies. Same for config generation in EDS policy (TODO).

Previously generating service config with CDS LB config very simple (a raw JSON string, only ~10 lines). But now, it's bad to write all of this here (config generation in EDS policy is probably to be even more complicated). This is unreadable, with three levels of child policy nesting (maps with values of maps with values of maps). We would need to seek for a better approach:

Similar to how LB config is parsed (each LB config parser recursively calls child policy's own parser to parse child policy), we can implement a LB config generator for each LB policy so that each config generator generates its own config (recursively use child policy's generator to generate child policies) and put the output together.

Reply:
Reorganized code. Seems a little better.